util/filesort/filesort.go

```diff
@@ -0,0 +1,490 @@
+	cursor    int                        // required when performing full in-memory sort
+	closed    bool
+	fetched   bool
+	rowHeap   *rowHeap
+	fds       []*os.File
+	fileCount int
+	err       error
+}
+
+// Builder builds a new FileSorter.
+type Builder struct {
+	sc      *variable.StatementContext
+	keySize int
+	valSize int
+	bufSize int
+	byDesc  []bool
+	tmpDir  string
+}
+
+// SetSC sets StatementContext instance which is required in row comparison.
+func (b *Builder) SetSC(sc *variable.StatementContext) *Builder {
+	b.sc = sc
+	return b
+}
+
+// SetSchema sets the schema of row, including key size and value size.
+func (b *Builder) SetSchema(keySize, valSize int) *Builder {
+	b.keySize = keySize
+	b.valSize = valSize
+	return b
+}
+
+// SetBuf sets the number of rows FileSorter can hold in memory at a time.
+func (b *Builder) SetBuf(bufSize int) *Builder {
+	b.bufSize = bufSize
+	return b
+}
+
+// SetDesc sets the ordering rule of row comparison.
+func (b *Builder) SetDesc(byDesc []bool) *Builder {
+	b.byDesc = byDesc
+	return b
+}
+
+// SetDir sets the working directory for FileSorter.
+func (b *Builder) SetDir(tmpDir string) *Builder {
+	b.tmpDir = tmpDir
+	return b
+}
+
+// Build creates a FileSorter instance using given data.
+func (b *Builder) Build() (*FileSorter, error) {
+	// Sanity checks
+	if b.sc == nil {
+		return nil, errors.New("StatementContext is nil")
+	}
+	if b.keySize != len(b.byDesc) {
+		return nil, errors.New("mismatch in key size and byDesc slice")
+	}
+	if b.keySize <= 0 {
+		return nil, errors.New("key size is not positive")
+	}
+	if b.valSize <= 0 {
+		return nil, errors.New("value size is not positive")
+	}
+	if b.bufSize <= 0 {
+		return nil, errors.New("buffer size is not positive")
+	}
+	_, err := os.Stat(b.tmpDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errors.New("tmpDir does not exist")
+		}
+		return nil, errors.Trace(err)
+	}
+
+	rh := &rowHeap{sc: b.sc,
+		ims:    make([]*item, 0),
+		byDesc: b.byDesc,
+		err:    nil,
+	}
+
+	return &FileSorter{sc: b.sc,
+		keySize:   b.keySize,
+		valSize:   b.valSize,
+		bufSize:   b.bufSize,
+		buf:       make([]*comparableRow, 0, b.bufSize),
+		files:     make([]string, 0),
+		byDesc:    b.byDesc,
+		cursor:    0,
+		fetched:   false,
+		closed:    false,
+		rowHeap:   rh,
+		tmpDir:    b.tmpDir,
+		fds:       make([]*os.File, 0),
+		fileCount: 0,
+		err:       nil,
+	}, nil
+}
+
+func (fs *FileSorter) Len() int { return len(fs.buf) }
+
+func (fs *FileSorter) Swap(i, j int) { fs.buf[i], fs.buf[j] = fs.buf[j], fs.buf[i] }
+
+func (fs *FileSorter) Less(i, j int) bool {
+	l := fs.buf[i].key
+	r := fs.buf[j].key
+	ret, err := lessThan(fs.sc, l, r, fs.byDesc)
+	if fs.err == nil {
+		fs.err = err
+	}
+	return ret
+}
+
+func (fs *FileSorter) uniqueFileName() string {
+	ret := path.Join(fs.tmpDir, strconv.Itoa(fs.fileCount))
+	fs.fileCount++
+	return ret
+}
+
+func (fs *FileSorter) flushMemory() error {
```

Comment:
Added comment. Changed function name to `flushMemory` to `flushToFile`, a more readable name.